connector/servicegraphconnector/connector_test.go

```diff
@@ -436,3 +450,122 @@ func TestValidateOwnTelemetry(t *testing.T) {
 	}
 	metricdatatest.AssertEqual(t, want, got, metricdatatest.IgnoreTimestamp())
 }
+
+func TestExtraDimensionsLabels(t *testing.T) {
+	extraDimensions := []string{"db.system", "messaging.system"}
+	cfg := &Config{
+		Dimensions:              extraDimensions,
+		LatencyHistogramBuckets: []time.Duration{time.Duration(0.1 * float64(time.Second)), time.Duration(1 * float64(time.Second)), time.Duration(10 * float64(time.Second))},
+		Store:                   StoreConfig{MaxItems: 10},
+	}
+
+	set := componenttest.NewNopTelemetrySettings()
+	set.Logger = zaptest.NewLogger(t)
+	conn, err := newConnector(set, cfg, newMockMetricsExporter())
+	assert.NoError(t, err)
+
+	assert.NoError(t, conn.Start(context.Background(), componenttest.NewNopHost()))
+	defer require.NoError(t, conn.Shutdown(context.Background()))
+
+	td, err := golden.ReadTraces("testdata/extra-dimensions-queue-db-trace.yaml")
+	assert.NoError(t, err)
+	assert.NoError(t, conn.ConsumeTraces(context.Background(), td))
+
+	conn.store.Expire()
+
+	metrics := conn.metricsConsumer.(*mockMetricsExporter).GetMetrics()
+	require.Len(t, metrics, 1)
+
+	expectedMetrics, err := golden.ReadMetrics("testdata/extra-dimensions-queue-db-expected-metrics.yaml")
+	assert.NoError(t, err)
+
+	err = pmetrictest.CompareMetrics(expectedMetrics, metrics[0],
+		pmetrictest.IgnoreStartTimestamp(),
+		pmetrictest.IgnoreTimestamp(),
+	)
+	require.NoError(t, err)
+}
+
+func TestVirtualNodeServerLabels(t *testing.T) {
+	virtualNodeDimensions := []string{"peer.service", "db.system", "messaging.system"}
+	cfg := &Config{
+		Dimensions:                virtualNodeDimensions,
+		LatencyHistogramBuckets:   []time.Duration{time.Duration(0.1 * float64(time.Second)), time.Duration(1 * float64(time.Second)), time.Duration(10 * float64(time.Second))},
+		Store:                     StoreConfig{MaxItems: 10},
+		VirtualNodePeerAttributes: virtualNodeDimensions,
+		VirtualNodeExtraLabel:     true,
+		MetricsFlushInterval:      time.Millisecond,
+	}
+
+	set := componenttest.NewNopTelemetrySettings()
+	set.Logger = zaptest.NewLogger(t)
+
+	trace := "testdata/virtual-node-label-server-trace.yaml"
+	expected := "testdata/virtual-node-label-server-expected-metrics.yaml"
+
+	conn, err := newConnector(set, cfg, newMockMetricsExporter())
+	assert.NoError(t, err)
+	assert.NoError(t, conn.Start(context.Background(), componenttest.NewNopHost()))
+
+	td, err := golden.ReadTraces(trace)
+	assert.NoError(t, err)
+	assert.NoError(t, conn.ConsumeTraces(context.Background(), td))
+
+	conn.store.Expire()
+	time.Sleep(2 * time.Millisecond) // Wait for metrics to be flushed
```

Comment:
can you check that metrics have been flushed in some other way? if so, you can use assert.Eventually, avoiding waiting a specific amount of time, which is a common source of flakyness in tests

Reply:
That's a great idea - added. Thanks!